File: frontend/lib/types/driver.ts
export interface DriverCareerStats {
  total_race_starts: number;
  total_wins: number;
  total_podiums: number;
  total_pole_positions: number;
  total_fastest_laps: number;
  total_championship_points: number;
  championships_won: number;
  career_win_rate: number;
  career_podium_rate: number;
  dnf_rate: number;
  avg_qualifying_position: number | null;
  avg_finish_position: number | null;
  seasons_active: number[];
  teams_driven_for: string[];
  active_status: boolean;
  last_race_date: string | null;
}

export interface DriverSummaryOut {
  id: string;
  full_name: string;
  nationality: string;
  active_status: boolean;
  career_stats: DriverCareerStats;
}

export interface DriverSeason {
  season_year: number;
  championship_position: number | null;
  points: number;
  wins: number;
  podiums: number;
  poles: number;
  fastest_laps: number;
  races_entered: number;
  dnfs: number;
  team_name?: string;
  teammate_qualifying_delta: number | null;
  teammate_race_delta: number | null;
}

export interface DriverOut extends DriverSummaryOut {
  first_name: string;
  last_name: string;
  date_of_birth: string | null;
  biography: string | null;
  profile_image_url: string | null;
  license_grade: string | null;
  ergast_driver_id: string | null;
  seasons: DriverSeason[];
  created_at: string;
  updated_at: string;
}

export interface DriverListResponse {
  total: number;
  page: number;
  limit: number;
  drivers: DriverSummaryOut[];
}

export type SortBy =
  | "win_rate"
  | "total_wins"
  | "total_podiums"
  | "championships_won";

export interface DriversQueryParams {
  nationality?: string;
  min_age?: number;
  max_age?: number;
  min_wins?: number;
  max_wins?: number;
  min_podiums?: number;
  min_championships?: number;
  active_only?: boolean;
  sort_by?: SortBy;
  sort_order?: "asc" | "desc";
  page?: number;
  limit?: number;
}

export interface DriverFilterRequest extends DriversQueryParams {}

File: app/schemas/driver.py
import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


class DriverCareerStatsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    total_race_starts: int
    total_wins: int
    total_podiums: int
    total_pole_positions: int
    total_fastest_laps: int
    total_championship_points: float
    championships_won: int
    career_win_rate: float
    career_podium_rate: float
    dnf_rate: float
    avg_qualifying_position: Optional[float]
    avg_finish_position: Optional[float]
    seasons_active: Optional[list[int]]
    teams_driven_for: Optional[list[str]]
    active_status: bool
    last_race_date: Optional[date]


class DriverOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    first_name: str
    last_name: str
    full_name: str
    nationality: Optional[str]
    date_of_birth: Optional[date]
    biography: Optional[str]
    profile_image_url: Optional[str]
    license_grade: Optional[str]
    active_status: bool
    career_stats: Optional[DriverCareerStatsOut]
    created_at: datetime
    updated_at: datetime


class DriverSummaryOut(BaseModel):
    """Lighter response for list endpoints."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    full_name: str
    nationality: Optional[str]
    active_status: bool
    career_stats: Optional[DriverCareerStatsOut]


class DriverSeasonOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    season_year: int
    championship_position: Optional[int]
    points: float
    wins: int
    podiums: int
    poles: int
    fastest_laps: int
    races_entered: int
    dnfs: int


class DriverListResponse(BaseModel):
    total: int
    page: int
    limit: int
    drivers: list[DriverSummaryOut]


class DriverCompareResponse(BaseModel):
    drivers: list[DriverOut]
